source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRTargetBuilderDirector.mm

```diff
@@ -0,0 +1,268 @@
+    }
+    return nil;
+}
+
+- (NSObject *)build:(std::shared_ptr<BaseActionElement> const &)action
+           director:(ACRTargetBuilderDirector *)director
+          ForButton:(UIButton *)button {
+    NSObject *target = [self build:action director:director];
+    if (target) {
+        [button addTarget:target
+                      action:@selector(doSelectAction)
+            forControlEvents:UIControlEventTouchUpInside];
+    }
+    return target;
+}
+
+@end
+
+@implementation ACRTargetBuilderDirector {
+    NSArray<NSDictionary<NSNumber *, ACRTargetBuilder *> *> *_builders;
+}
+
+- (instancetype)init {
+    self = [self init:nil capability:ACRAction adaptiveHostConfig:nil];
+    return self;
+}
+
+- (instancetype)init:(ACRView *)rootView
+            capability:(ACRTargetCapability)capability
+    adaptiveHostConfig:(ACOHostConfig *)adaptiveHostConfig {
+    self = [super init];
+    if (self) {
+        self.builderCapability = capability;
+        self.rootView = rootView;
+        self.adaptiveHostConfig = adaptiveHostConfig;
+
+        NSNumber *openUrl = [NSNumber numberWithInt:static_cast<int>(ActionType::OpenUrl)];
+        NSNumber *submit = [NSNumber numberWithInt:static_cast<int>(ActionType::Submit)];
+        NSNumber *showcard = [NSNumber numberWithInt:static_cast<int>(ActionType::ShowCard)];
+        NSNumber *toggle = [NSNumber numberWithInt:static_cast<int>(ActionType::ToggleVisibility)];
+
+        // target capabilities list supported events and corresponding target builders
+        _builders = @[
+            // ACRAction
+            @{
+                openUrl : [ACRAggregateTargetBuilder getInstance],
+                submit : [ACRAggregateTargetBuilder getInstance],
+                showcard : [ACRShowCardTargetBuilder getInstance],
+                toggle : [ACRToggleVisibilityTargetBuilder getInstance]
+            },
+            // ACRSelectAction
+            @{
+                openUrl : [ACRAggregateTargetBuilder getInstance],
+                submit : [ACRAggregateTargetBuilder getInstance],
+                toggle : [ACRToggleVisibilityTargetBuilder getInstance]
+            },
+            // ACRQuickReply
+            @{
+                openUrl : [ACRAggregateTargetBuilder getInstance],
+                submit : [ACRAggregateTargetBuilder getInstance],
+                toggle : [ACRToggleVisibilityTargetBuilder getInstance]
+            }
+        ];
+    }
+    return self;
+}
+
+- (NSObject *)build:(std::shared_ptr<BaseActionElement> const &)action {
+    if ([self checkAgainstMyCapability:action]) {
+        // check fail, stop and return
+        return nil;
+    }
+
+    ACRTargetBuilder *builder = [self getBuilder:action];
+    return [builder build:action director:self];
+}
+
+- (NSObject *)build:(std::shared_ptr<BaseActionElement> const &)action
+          forButton:(UIButton *)button {
+    if (![self checkAgainstMyCapability:action]) {
```

Comment:
>![self checkAgainstMyCapability:action] [](start = 8, length = 39)

This is a very confusing way of writing this logic. It looks like `checkAgainstMyCapability` is returning a boolean, but it's actually returning an enum where 0 (== ACROk) happens to be the success value.

Reply:
thanks; 

---
In reply to: [321507661](https://github.com/microsoft/AdaptiveCards/pull/3354#discussion_r321507661) [](ancestors = 321507661)